Add new `Rails/SquishedSQLHeredocs` cop

mobilutz opened the pull request:
This cops check `<<-SQL` SQL-heredocs if they use `.squish` or not.

Base is this styleguide: [Squished Heredocs](https://rails.rubystyle.guide/#squished-heredocs)

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [ ] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](https://github.com/rubocop-hq/rubocop-rails/blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop-rails/blob/master/CONTRIBUTING.md#changelog-entry-format).
* [ ] If this is a new cop, consider making a corresponding update to the [Rails Style Guide](https://github.com/rubocop-hq/rails-style-guide).
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

eugeneius commented on config/default.yml:
```diff
@@ -610,6 +610,13 @@ Rails/SkipsModelValidations:
+Rails/SquishedHeredocs:
+  Description: 'Checks the usage of `.squish` on SQL heredocs.'
+  Reference: 'https://rails.rubystyle.guide/#squished-heredocs'
+  Enabled: true
```
New cops should now be pending:

https://docs.rubocop.org/rubocop/versioning.html#pending-cops

```suggestion
  Enabled: 'pending'
```

eugeneius commented on config/default.yml:
```diff
@@ -610,6 +610,13 @@ Rails/SkipsModelValidations:
+  Reference: 'https://rails.rubystyle.guide/#squished-heredocs'
+  Enabled: true
+  VersionAdded: '2.7'
+  VersionChanged: '2.7'
```
I don't think `VersionChanged` should be set for new cops.

eugeneius commented on lib/rubocop/cop/rails/squished_heredocs.rb:
```diff
@@ -0,0 +1,80 @@
+  module Cop
+    module Rails
+      #
+      # Checks SQL heredocs to use `.squish`
```
I think we should include some detail here on why using `squish` is a good idea.

Since this cop only applies to SQL heredocs, the name should probably reflect that too (`SquishedSQLHeredocs`)?

mobilutz replied:
@eugeneius I changed the name of the cop. 

Regarding the more details, isn't it enough to link to the rails-styleguide?

eugeneius replied:
I checked some other cops with style guide references, and many had no justification in their documentation either. 👍

eugeneius commented on lib/rubocop/cop/rails/squished_heredocs.rb:
```diff
@@ -0,0 +1,80 @@
+          no_sql_heredoc?(node) || using_squish_on_sql_heredoc?(node)
+        end
+
+        def no_sql_heredoc?(node)
+          delimiter_string(node) != SQL
+        end
```
I'd invert this method for clarity, probably `no_offense_detected?` too.

```suggestion
        def sql_heredoc?(node)
          delimiter_string(node) == SQL
        end
```

eugeneius commented on lib/rubocop/cop/rails/squished_heredocs.rb:
```diff
@@ -0,0 +1,80 @@
+          range_by_whole_lines(range_between(pos, pos))
+        end
+
+        # Returns '~', '-' or nil
+        def heredoc_indent_type(node)
+          node.source[/^<<([~-])/, 1]
+        end
```
This looks unused.

eugeneius commented on spec/rubocop/cop/rails/squished_heredocs_spec.rb:
```diff
@@ -0,0 +1,37 @@
+
+  it 'registers an offense' do
+    expect_offense(<<~RUBY)
+      <<~SQL
+      ^^^^^^ Use `<<~SQL.squish` instead of `<<~SQL`.
+        SELECT * FROM table;
+      SQL
```
I'm not sure there should be an offense here: it's a single line squiggly heredoc, `squish` will have no effect.

I'd even say that ideally, this cop would remove `squish` from single line heredocs, converting them to be squiggly if necessary.

mobilutz replied:
Done

eugeneius commented on spec/rubocop/cop/rails/squished_heredocs_spec.rb:
```diff
@@ -0,0 +1,37 @@
+      ^^^^^^ Use `<<~SQL.squish` instead of `<<~SQL`.
+        SELECT * FROM table;
+      SQL
+    RUBY
```
All of the tests here inspect a heredoc on its own, but it would be good to also cover a heredoc defined as an inline method argument, and in particular when it's followed by another argument:

```ruby
execute(<<~SQL, "Post Load")
  SELECT * FROM posts
    WHERE post_id = 1
SQL

mobilutz replied:
This is a very very good point, which I did not think about before.

I will look into it, it will make the development for the cop a bit more complex for me 😉

mobilutz replied:
I did spend some time now, trying to find a solution for the `execute(<<~SQL, "Post Load")` code, but didn't find on yet.

I saw, that the base cop [`Layout/HeredocArgumentClosingParenthesis`[(https://github.com/rubocop-hq/rubocop/blob/master/lib/rubocop/cop/layout/heredoc_argument_closing_parenthesis.rb) has some logic around handling these types of heredocs, but looking at the code - it is quite complex and I guess grew over time as well.

I did add a check, to not check Heredocs that are used directly as a method argument, although the check might not be 100% correct I bet :|

@eugeneius What do you say to the newest implementation?

eugeneius replied:
I think `Layout/HeredocArgumentClosingParenthesis` is tricky because it needs to move code around the heredoc body. This cop doesn't need to modify the body, so it shouldn't be as difficult to handle all heredocs. You are right that it's a bit more complex, though 🙂 

The new approach seems fine to me, although I think there should be a test to demonstrate that it doesn't mangle the code I posted above. You could use `expect_no_offenses` to do that.

eugeneius commented:
Sorry @mobilutz, I just realised that my earlier suggestion about single-line heredocs was wrong. Even when a heredoc is only a single line, it still always has a trailing newline character:

```bash
$ ruby -e "puts <<~SQL.inspect
SELECT * FROM posts
SQL"
"SELECT * FROM posts\n"
```

Since the point of the style guide entry is to avoid logging newline characters, I think we should actually enforce `squish` in all cases, as you had originally. 😖

mobilutz commented:
> Since the point of the style guide entry is to avoid logging newline characters, I think we should actually enforce `squish` in all cases, as you had originally. 😖

Thanks for testing it out again, I totally did not do that.
I will revert the single line check so that `squish` will be used on all heredocs.

mobilutz commented:
@eugeneius Any chance you could review this again?
Thanks

eugeneius commented on lib/rubocop/cop/rails/squished_sql_heredocs.rb:
```diff
@@ -0,0 +1,92 @@
+
+        def autocorrect(node)
+          lambda do |corrector|
+            corrector.insert_after(heredoc_begin_line_range(node), SQUISH)
```
```suggestion
            corrector.insert_after(node, SQUISH)
```

eugeneius commented on lib/rubocop/cop/rails/squished_sql_heredocs.rb:
```diff
@@ -0,0 +1,92 @@
+        private
+
+        def offense_detected?(node)
+          inline_heredoc?(node) && sql_heredoc?(node) && offense?(node)
```
```suggestion
          sql_heredoc?(node) && offense?(node)
```

eugeneius commented on lib/rubocop/cop/rails/squished_sql_heredocs.rb:
```diff
@@ -0,0 +1,92 @@
+        end
+
+        def using_squish_on_sql_heredoc?(node)
+          heredoc_opening(node).end_with?(SQL_SQUISH)
```
```suggestion
          node.parent&.send_type? && node.parent.method?(:squish)
```

eugeneius commented on lib/rubocop/cop/rails/squished_sql_heredocs.rb:
```diff
@@ -0,0 +1,92 @@
+          format(
+            MSG,
+            expect: "#{heredoc_type(node)}#{SQL_SQUISH}",
+            current: heredoc_opening(node)
```
```suggestion
            current: node.source
```

mobilutz commented:
@eugeneius Thank you very much for the great input! Now even method-parameter Heredocs are working!!!

I rebased `origin/master` and added hopefully all documentation etc.
Is there anything else I can do to make this PR mergeable?

eugeneius commented on config/default.yml:
```diff
@@ -638,6 +638,12 @@ Rails/SkipsModelValidations:
+  Description: 'Checks multi- and singleline SQL heredocs to use `.squish`.'
+  StyleGuide: 'https://rails.rubystyle.guide/#squished-heredocs'
+  Enabled: 'pending'
+  VersionAdded: '2.7'
```
2.7 has already been released, this should be 2.8 now.

eugeneius commented on lib/rubocop/cop/rails/squished_sql_heredocs.rb:
```diff
@@ -0,0 +1,89 @@
+  module Cop
+    module Rails
+      #
+      # Checks multi- and singleline SQL heredocs to use `.squish`.
```
The cop doesn't make any distinction between multi- and single-line heredocs any more, do we need to mention them here?

eugeneius commented on lib/rubocop/cop/rails/squished_sql_heredocs.rb:
```diff
@@ -0,0 +1,89 @@
+      #
+      class SquishedSQLHeredocs < Cop
+        include Heredoc
+        include RangeHelp
```
This is no longer used ✂️

eugeneius commented on lib/rubocop/cop/rails/squished_sql_heredocs.rb:
```diff
@@ -0,0 +1,89 @@
+        private
+
+        def offense_detected?(node)
+          sql_heredoc?(node) && offense?(node)
```
I'd rename `using_squish_on_sql_heredoc?` to `using_squish?`, call it here directly, and remove `offense?`:

```suggestion
          sql_heredoc?(node) && !using_squish?(node)
```

"SQL heredoc and not using squish" makes the logic pretty clear 🙂

eugeneius commented on lib/rubocop/cop/rails/squished_sql_heredocs.rb:
```diff
@@ -0,0 +1,89 @@
+        def message(node)
+          format(
+            MSG,
+            expect: "#{heredoc_type(node)}#{SQL_SQUISH}",
```
This seems simpler as:

```suggestion
            expect: "#{node.source}#{SQUISH}",
```

That would also make `SQL_SQUISH` unnecessary.

koic commented:
@mobilutz Can you update this PR?

koic commented:
Thank you!
